File: app/src/main/cpp/SampleJni.c
//
// Created by akuang on 2/11/2018.
//

#include <jni.h>

// TODO:
// From https://docs.oracle.com/javase/1.5.0/docs/guide/jni/spec/invocation.html:
// The JNI interface pointer (JNIEnv) is valid only in the current thread. Should another thread
// need to access the Java VM, it must first call AttachCurrentThread() to attach itself to the
// VM and obtain a JNI interface pointer. Once attached to the VM, a native thread works just like
// an ordinary Java thread running inside a native method. The native thread remains attached to
// the VM until it calls DetachCurrentThread() to detach itself.

jstring
Java_com_example_akuang_androidjni_SampleJni_nativeStringFromJNI(JNIEnv *env, jobject obj) {
    char* hello = "Hello from C";
    return (*env)->NewStringUTF(env, hello);
}

jstring
Java_com_example_akuang_androidjni_SampleJni_nativeStringFromJAVA(JNIEnv *env, jobject obj) {

    // Example of calling a Java method from C++

    jclass sampleJinClass = (*env)->GetObjectClass(env, obj);

    if (sampleJinClass == 0) {
        return (*env)->NewStringUTF(env, "[SampleJni.c] GetObjectClass failed!");
    }

    jfieldID javaField = (*env)->GetFieldID(env, sampleJinClass, "javaString", "Ljava/lang/String;");

    if (javaField == 0) {
        return (*env)->NewStringUTF(env, "[SampleJni.c] GetFieldID failed!");
    }

    jstring javaString = (jstring)(*env)->GetObjectField(env, obj, javaField);

    if (javaString == NULL) {
        return (*env)->NewStringUTF(env, "[SampleJni.c] GetObjectField failed!");
    }

    const char* cstr = (*env)->GetStringUTFChars(env, javaString, NULL);

    return (*env)->NewStringUTF(env, cstr);
}

File: app/src/main/java/com/example/akuang/androidjni/SampleJni.java
package com.example.akuang.androidjni;

/**
 * Created by akuang on 2/11/2018.
 */

public class SampleJni {

    // Sample field for calling Java from C
    private String javaString = "Hello from Java field!";

    // Sample method calling C method which calls back to Java field
    public String stringFromJava()
    {
        return nativeStringFromJAVA();
    }

    // Sample method calling C from Java
    public String stringFromJni()
    {
        return nativeStringFromJNI();
    }

    /**
     * A native method that is implemented by the 'native-lib' native library,
     * which is packaged with this application.
     */
    private native String nativeStringFromJNI();
    private native String nativeStringFromJAVA();
}
